Add WAVEFORMATEX format to WavWriter

DenisKolodin opened the pull request:
Traditional wav format is the only one supported format
for some musical devices (Akai MPC for example) and some DAWs.

ruuda commented:
Thank you for taking the time to implement this. I will be offline for a few days, so I will take a closer look next week.

DenisKolodin commented:
Sure thing! Feel free to request any update for this.

ruuda commented on src/write.rs:
```diff
@@ -176,6 +179,13 @@ impl<W> WavWriter<W>
+            // which includes the WAVEFORMAT structure.
+            extensible: spec.channels > 2 ||
+                        spec.bits_per_sample > 16 ||
+                        spec.sample_format == SampleFormat::Float,
```
The `write_waveformatex` method below does handle a float sample format, this check is not required here, is it?

ruuda commented on src/write.rs:
```diff
@@ -187,9 +197,76 @@ impl<W> WavWriter<W>
+
+        // Write the header in-memory first.
+        {
+            let mut buffer: io::Cursor<&mut [u8]> = io::Cursor::new(&mut header);
```
As `WAVEFORMATEXTENSIBLE` includes a `WAVEFORMATEX`, we could extract writing this part into a method, so it could be used to write both headers.

ruuda commented on src/write.rs:
```diff
@@ -341,7 +418,12 @@ impl<W> WavWriter<W>
         let file_size = self.data_bytes_written + 60;
```
This is no longer the case now, the file size header should be updated as well.

DenisKolodin commented:
I added the necessary changes. Check, please, have I made it right?
It's a little tricky to write whole `WAVEFORMATEX` struct for both formats so I implemented some methods to write different parts of that struct.

DenisKolodin commented:
@ruuda The one test hasn't passed, because it couldn't bootstrap the building environment. What do you think about the changes? I can update it again if necessary.

ruuda commented:
Don’t worry about the Travis builds, it often does that for older Rust versions. Could be a deadlock in Cargo, but those older versions will not get patched any more anyway ...

Thanks a lot for doing this!

ruuda commented:
Hound 3.2.0, which includes this change, is now on crates.io.

DenisKolodin commented:
Cool! Thank you!
